### PR TITLE
Upgrade django version to 4.1

### LINKED ISSUE #271: Upgrade Django
We're on a really old version (2.2) and should upgrade. I haven't been using Django for many years, so would really appreciate some help here.  At the time of writing, 4.1 seems like the best choice, then we can bump to 4.2 when that comes out with hopefully not too much work.

See supported versions at https://www.djangoproject.com/download/#supported-versions

### PR DESCRIPTION
**Pull request by PatriotRossii**
closes #271

### CODE REVIEW

**Comment by tocic** [comment]
- Looks like [the minimum supported python version](https://docs.djangoproject.com/en/4.1/releases/4.0/#python-compatibility) is 3.8 now, not 3.6 specified in README and CI.
- Running `python -Wd manage.py test` produces `RemovedInDjango50Warning: PickleSerializer is deprecated due to its security risk. Use JSONSerializer instead`. Perhaps worth fixing now too.
- `pytz` requirement is [probably redundant now](https://docs.djangoproject.com/en/4.1/releases/4.0/#zoneinfo-default-timezone-implementation).

I haven't read all the release notes, so this list may not be exhaustive.

**Comment by PatriotRossii** [comment]
> * Looks like [the minimum supported python version](https://docs.djangoproject.com/en/4.1/releases/4.0/#python-compatibility) is 3.8 now, not 3.6 specified in README and CI.
> 
>     * Running `python -Wd manage.py test` produces `RemovedInDjango50Warning: PickleSerializer is deprecated due to its security risk. Use JSONSerializer instead`. Perhaps worth fixing now too.
> 
>     * `pytz` requirement is [probably redundant now](https://docs.djangoproject.com/en/4.1/releases/4.0/#zoneinfo-default-timezone-implementation).
> 
> 
> I haven't read all the release notes, so this list may not be exhaustive.

Switching from `PickleSerializer` to `JSONSerializer`  requires writing of custom JSONSerializer for a lot of classes. Honestly, I don't want to write it because I'm not familiar with the best practices how to write it well

I think `PickleSerializer` is fine for now, django 4.2 is a LTS version, so after we update to it there'll be more than two years to write that custom `JSONSerializer`

**Comment by knatten** [comment]
Thanks for the PR! That was a lot less work than I feared. I don't know Django very well, and just remembered how much work it was to upgrade from 1 to 2. Seems like they made a lot fewer big changes in 3 and 4.

I'm ok with dropping support for Python 3.6 and 3.7, and will see about upgrading Python on the web host.

File: requirements.frozen.txt [hunk 1]
```diff
@@ -1,18 +1,19 @@
+asgiref==3.6.0
 autopep8==2.0.1
 certifi==2022.12.7
 charset-normalizer==2.0.12
-Django==2.2.28
+Django==4.1
```
Comment (knatten) [inline requirements.frozen.txt:5]: We also need to bump Django `requirements.txt`. The way I've been upgrading pip dependencies until now has been to bump `requirements.txt`, creating a fresh venv, installing `requirements.txt` and then do `pip freeze > requirements.frozen.txt`, but I've been meaning to improve on that. Created https://github.com/knatten/cppquiz/issues/281 for it now.

Whether we use the old method in this PR or solve #281 first is both fine with me.

File: README.md [hunk 2]
```diff
@@ -6,7 +6,7 @@ http://creativecommons.org/licenses/by-sa/4.0/
 ![Build Status](https://github.com/knatten/cppquiz/actions/workflows/ci.yml/badge.svg)
 
 # Requirements
-- Python 3.6 or 3.7
+- Python 3.8 / python 3.9 / python 3.10
```
Comment (knatten) [inline README.md:9]: I suggest just saying "Python 3.8 or higher" here, then we don't need to keep it in sync with ci.yml.

**Comment by knatten** [comment]
Great PR! Very nice with the small commits with good explanations too, made it easy to review. I just have a few comments/questions.

File: cppquiz/settings.py [hunk 3]
```diff
@@ -37,10 +37,6 @@
 # to load the internationalization machinery.
 USE_I18N = True
 
-# If you set this to False, Django will not format dates, numbers and
-# calendars according to the current locale
-USE_L10N = True
-
```
[the remark this answers is not in the document]
Comment (PatriotRossii) [reply cppquiz/settings.py:43]: No, it won't

It's default value is True from Django 4.0 ([source](https://docs.djangoproject.com/en/4.1/ref/settings/#std-setting-USE_L10N))

File: requirements.frozen.txt (hunk 1 above)
Comment (PatriotRossii) [reply requirements.frozen.txt:5]: https://github.com/knatten/cppquiz/pull/279/commits/d05a4d379b2ab519c157099ba3a52c6eaf1ea08a

File: README.md (hunk 2 above)
Comment (PatriotRossii) [reply README.md]: https://github.com/knatten/cppquiz/pull/279/commits/2a546f5d7e1b30659b38b2f9ecbec2bf4fd27df0

File: cppquiz/settings.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (knatten) [reply cppquiz/settings.py:43]: Right, nice!

**Comment by knatten** [comment]
> > * Looks like [the minimum supported python version](https://docs.djangoproject.com/en/4.1/releases/4.0/#python-compatibility) is 3.8 now, not 3.6 specified in README and CI.
> >   
> >   * Running `python -Wd manage.py test` produces `RemovedInDjango50Warning: PickleSerializer is deprecated due to its security risk. Use JSONSerializer instead`. Perhaps worth fixing now too.
> >   * `pytz` requirement is [probably redundant now](https://docs.djangoproject.com/en/4.1/releases/4.0/#zoneinfo-default-timezone-implementation).
> > 
> > I haven't read all the release notes, so this list may not be exhaustive.
> 
> Switching from `PickleSerializer` to `JSONSerializer` requires writing of custom JSONSerializer for a lot of classes. Honestly, I don't want to write it because I'm not familiar with the best practices how to write it well
> 
> I think `PickleSerializer` is fine for now, django 4.2 is a LTS version, so after we update to it there'll be more than two years to write that custom `JSONSerializer`

Yeah we don't need to fix that in this PR, but would be good to fix at some point.

**Comment by knatten** [comment]
I've installed Python 3.11 and deployed this branch to https://beta.cppquiz.org now. Seems to be working well!

**Comment by knatten** [comment]
This has now been deployed to the main site. Thanks a lot for fixing!

**Pull request merged**